packages/core/src/indexing/insertion-order.ts

```diff
@@ -28,45 +36,69 @@ type QueryFunc = (bldr: Knex.QueryBuilder<any, any>) => Knex.QueryBuilder<any, a
 abstract class Cursor {
   /**
    * Decode cursor from base64url as JSON.
+   * Return `undefined` if +cursor+ is `undefined` or `null`.
    */
-  static parse(cursor: string): any {
-    return JSON.parse(uint8arrays.toString(uint8arrays.fromString(cursor, 'base64url')))
+  static parse<Content extends StreamContent = StreamContent>(
+    cursor: string | undefined
+  ): CursorData<Content> {
+    return cursor
+      ? JSON.parse(uint8arrays.toString(uint8arrays.fromString(cursor, 'base64url')))
+      : undefined
   }
 
   /**
    * base64url-encode cursor from +input+ object.
    * Return `undefined` if +input+ object is `undefined` or `null`.
    */
-  static stringify(input: any): string | undefined {
-    if (input) {
-      return uint8arrays.toString(uint8arrays.fromString(JSON.stringify(input)), 'base64url')
-    } else {
+  static stringify(
+    input: QueryResult | undefined,
+    orderByKeys: Array<string> = []
+  ): string | undefined {
+    if (input == null) {
       return undefined
     }
-  }
-}
 
-/**
- * Prepare insertion cursor.
- */
-function asInsertionCursor(input: { created_at: number } | undefined) {
-  if (!input) return undefined
-  return { created_at: input.created_at }
+    let cursor: CursorData
+    if (orderByKeys.length === 0) {
+      // Use `created_at` field
+      cursor = { type: 'timestamp', id: input.stream_id, value: input.created_at }
+    } else {
+      // Use custom content fields
+      const content = JSON.parse(input.stream_content)
+      cursor = orderByKeys.reduce(
```

Comment:
NITPICK, feel free to ignore:

is this code equivalent to:
```
cursor = { type: 'content', id: input.stream_id, value: {} }
for (const key of orderByKeys) {
  if (content[key] != null) {
    cursor.value[key] = content[key]
  }
}
```
?

If so, I'd personally prefer the for-loop version as I think it's much more immediately obvious to a reader what is happening.  I generally try to err on the side of verbosity and straightforwardness over anything that comes close to "magic", even at the cost of "elegance".  
The reduce function is standard and well-understood enough to not really qualify as magic, so I think it's fine if you prefer to leave it as-written, but I at least had to think for a few seconds to understand what this was doing, and anytime a reader has to think for a few seconds to understand some code it is an increase in cognitive load for the reader and a signal that it should be simplified if possible, IMHO.

Reply:
Yes no problem changing to a for loop